akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,341 @@
+        poolState = Terminated
+        terminationCondition.signalAll()
+      }
+    }
+
+  override def shutdownNow(): util.List[Runnable] =
+    locked(bookKeepingLock) {
+      poolState = ShutDown
+      workers.foreach(_.stop())
+      attemptPoolTermination()
+      Collections.emptyList[Runnable]() // like in the FJ executor, we do not provide facility to obtain tasks that were in queue
+    }
+
+  override def shutdown(): Unit =
+    locked(bookKeepingLock) {
+      poolState = ShuttingDown
+      // interrupts only idle workers.. so others can process their queues
+      workers.foreach(_.stopIfIdle())
+      attemptPoolTermination()
+    }
+
+  override def isShutdown: Boolean = poolState == ShutDown
+
+  override def isTerminated: Boolean = poolState == Terminated
+
+  private class ThreadPoolWorker(val q: BoundedTaskQueue) extends Runnable {
+
+    sealed trait WorkerState
+    case object NotStarted extends WorkerState
+    case object InExecution extends WorkerState
+    case object Idle extends WorkerState
+
+    val thread: Thread = tf.newThread(this)
+    @volatile var workerState: WorkerState = NotStarted
+
+    def startWorker() = {
+      workerState = Idle
+      thread.start()
+    }
+
+    def runCommand(command: Runnable) = {
+      workerState = InExecution
+      try
+        command.run()
+      finally
+        workerState = Idle
+    }
+
+    override def run(): Unit = {
+
+      /**
+       * Determines whether the worker can keep running or not.
+       * In order to continue polling for tasks three conditions
+       * need to be satisfied:
+       *
+       * 1) pool state is less than Shutting down or queue
+       * is not empty (e.g pool state is ShuttingDown but there are still messages to process)
+       *
+       * 2) the thread backing up  this worker has not been interrupted
+       *
+       * 3) We are not in ShutDown state (in which we should not be processing any enqueued tasks)
+       */
+      def shouldKeepRunning =
+        (poolState < ShuttingDown || !q.isEmpty) &&
+          !Thread.interrupted() &&
+          poolState != ShutDown
+
+      var abruptTermination = true
+      try {
+        while (shouldKeepRunning) {
+          val c = q.poll()
+          if (c != null)
+            runCommand(c)
+          else // if not wait for a bit
+            waitingStrategy()
+        }
+        abruptTermination = false // if we have reached here, our termination is not due to an exception
+      } finally {
+        onWorkerExit(this, abruptTermination)
+      }
+    }
+
+    def stop() =
+      if (!thread.isInterrupted && workerState != NotStarted)
+        thread.interrupt()
+
+    def stopIfIdle() =
+      if (workerState == Idle)
+        stop()
+  }
+
+}
+
+object AffinityPool {
+
+  type WaitingStrategy = () ⇒ Unit
+  val busySpinWaitingStrategy = () ⇒ ()
+  val sleepWaitingStrategy = () ⇒ LockSupport.parkNanos(1l)
+  val yieldWaitingStrategy = () ⇒ Thread.`yield`()
```

Comment:
make as much as possible private or `private[akka]` (with the additional internal annotations)